File: shared/src/commonMain/kotlin/com/slowthemdown/shared/calculator/SpeedCalculator.kt
package com.slowthemdown.shared.calculator

import com.slowthemdown.shared.model.TrafficStats
import kotlin.math.sqrt

object SpeedCalculator {
    /**
     * Calculate speed in m/s from pixel displacement, pixels-per-meter, and time delta.
     * Formula: (pixelDisplacement / pixelsPerMeter) meters / timeDeltaSeconds seconds → m/s
     */
    fun calculateSpeed(
        pixelDisplacement: Double,
        pixelsPerMeter: Double,
        timeDeltaSeconds: Double
    ): Double {
        if (pixelsPerMeter <= 0 || timeDeltaSeconds <= 0) return 0.0
        val distanceMeters = pixelDisplacement / pixelsPerMeter
        return distanceMeters / timeDeltaSeconds
    }

    /** Calculate pixels per meter from a known reference distance in meters */
    fun pixelsPerMeter(pixelDistance: Double, referenceMeters: Double): Double {
        if (referenceMeters <= 0) return 0.0
        return pixelDistance / referenceMeters
    }

    /** Compute V85 — the interpolated 85th percentile speed */
    fun v85(speeds: List<Double>): Double? {
        if (speeds.isEmpty()) return null
        val sorted = speeds.sorted()
        val n = sorted.size.toDouble()
        val rank = 0.85 * (n - 1)
        val lower = rank.toInt()
        val upper = minOf(lower + 1, sorted.size - 1)
        val fraction = rank - lower.toDouble()
        return sorted[lower] + fraction * (sorted[upper] - sorted[lower])
    }

    /**
     * Compute traffic statistics for a set of speed/limit pairs.
     * Each pair is (speed in m/s, speedLimit in m/s).
     */
    fun trafficStats(entries: List<Pair<Double, Double>>): TrafficStats? {
        if (entries.isEmpty()) return null
        val speeds = entries.map { it.first }
        val sorted = speeds.sorted()
        val mean = speeds.sum() / speeds.size

        val overLimit = entries.count { it.first > it.second }
        val overPercent = overLimit.toDouble() / entries.size * 100

        return TrafficStats(
            count = entries.size,
            mean = mean,
            median = median(sorted),
            min = sorted.first(),
            max = sorted.last(),
            v85 = v85(speeds) ?: 0.0,
            overLimitCount = overLimit,
            overLimitPercent = overPercent,
            standardDeviation = standardDeviation(speeds, mean)
        )
    }

    private fun median(sorted: List<Double>): Double {
        if (sorted.isEmpty()) return 0.0
        val mid = sorted.size / 2
        return if (sorted.size % 2 == 0) {
            (sorted[mid - 1] + sorted[mid]) / 2
        } else {
            sorted[mid]
        }
    }

    private fun standardDeviation(values: List<Double>, mean: Double): Double {
        if (values.size <= 1) return 0.0
        val variance = values.sumOf { (it - mean) * (it - mean) } / (values.size - 1)
        return sqrt(variance)
    }
}

File: ios/SlowThemDown/Models/SpeedCalculator.swift
import Foundation

enum SpeedCalculator {
    /// Calculate speed in m/s from pixel displacement, pixels-per-meter, and time delta
    /// Formula: (pixelDisplacement / pixelsPerMeter) meters / timeDeltaSeconds seconds -> m/s
    static func calculateSpeed(
        pixelDisplacement: Double,
        pixelsPerMeter: Double,
        timeDeltaSeconds: Double
    ) -> Double {
        guard pixelsPerMeter > 0, timeDeltaSeconds > 0 else { return 0 }
        let distanceMeters = pixelDisplacement / pixelsPerMeter
        return distanceMeters / timeDeltaSeconds
    }

    /// Calculate pixels per meter from a known reference distance in meters
    static func pixelsPerMeter(pixelDistance: Double, referenceMeters: Double) -> Double {
        guard referenceMeters > 0 else { return 0 }
        return pixelDistance / referenceMeters
    }

    /// Compute V85 — the interpolated 85th percentile speed
    static func v85(speeds: [Double]) -> Double? {
        guard !speeds.isEmpty else { return nil }
        let sorted = speeds.sorted()
        let n = Double(sorted.count)
        let rank = 0.85 * (n - 1)
        let lower = Int(floor(rank))
        let upper = min(lower + 1, sorted.count - 1)
        let fraction = rank - Double(lower)
        return sorted[lower] + fraction * (sorted[upper] - sorted[lower])
    }

    /// Compute traffic statistics for a set of speed entries
    static func trafficStats(entries: [SpeedEntry]) -> TrafficStats? {
        guard !entries.isEmpty else { return nil }
        let speeds = entries.map(\.speed)
        let sorted = speeds.sorted()
        let mean = speeds.reduce(0, +) / Double(speeds.count)

        let overLimit = entries.filter(\.isOverLimit).count
        let overPercent = Double(overLimit) / Double(entries.count) * 100

        return TrafficStats(
            count: entries.count,
            mean: mean,
            median: median(sorted),
            min: sorted.first ?? 0,
            max: sorted.last ?? 0,
            v85: v85(speeds: speeds) ?? 0,
            overLimitCount: overLimit,
            overLimitPercent: overPercent,
            standardDeviation: standardDeviation(speeds, mean: mean)
        )
    }

    private static func median(_ sorted: [Double]) -> Double {
        guard !sorted.isEmpty else { return 0 }
        let mid = sorted.count / 2
        if sorted.count.isMultiple(of: 2) {
            return (sorted[mid - 1] + sorted[mid]) / 2
        }
        return sorted[mid]
    }

    private static func standardDeviation(_ values: [Double], mean: Double) -> Double {
        guard values.count > 1 else { return 0 }
        let variance = values.reduce(0) { $0 + ($1 - mean) * ($1 - mean) } / Double(values.count - 1)
        return sqrt(variance)
    }
}

struct TrafficStats {
    let count: Int
    let mean: Double
    let median: Double
    let min: Double
    let max: Double
    let v85: Double
    let overLimitCount: Int
    let overLimitPercent: Double
    let standardDeviation: Double
}
